### PR TITLE
Correction to CO2 and Generic Contaminant Modeling with Refrigeration Door Mixing

### LINKED ISSUE #8783: Unexpected CO2 and relative humidity results for mixing zones
Issue overview
--------------
An EnergyPlus model with mixing zone objects generated unexpected CO2 concentrations and relative humidity results. The model has four thermal zones and each zone has the same CO2 generation and infiltration rates. There are two non-mixing zones and two mixing zones. For the two mixing zones, each zone serves as both the "source" and "receiving" zone. With v9.5, CO2 concentrations for the two mixing zones vary with the design mixing flow rates. 
It is expected that CO2 concentration should not vary with the design mixing flows for the model. The CO2 mass balance equation (Eq.2.52) from Engineering Reference is listed below. Czi (CO2 concentration in the zone air being transferred into this zone)  and Czt (CO2 concentration in the zone air being transferred into this zone) should be the same in theory for both nearly identical mixing zones.

image

### Details
Some additional details for this issue (if relevant):
  - v9.5
  - ticket #15714

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file added EnergyPlusDevSupport\DefectFiles
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by RKStrand**
Pull request overview
---------------------
 - Fixes #8783 
 - This defect began at a look into some issues with mixing and CO2 level that a user noticed.  In a file with four identical zones, two of which had equal mixing with each other, the CO2 level for the two zones with mixing did not equal the CO2 level for the zones without mixing.  In theory, despite the mixing all zones should have the same CO2 level.  However, the problem is actually a function of the code because the CO2 level used for the "other" zone is the CO2 level from the previous time step.  So, with this lagged value, the CO2 levels of the zones with mixing will not track exactly with the zones without mixing.  The defect file used two time steps per hour and had a mixing ACH that was between 70-80 which made the differences much more noticeable.  With reasonable time steps and ACH values of around 4 showed some lagging, but nowhere near as much as in the defect file.  So, it was decided to not fix this since this is actually a result of the simulation design.  However, during analysis of this issue, a typo was discovered in the code for the modeling of generic contaminants when using refrigeration door mixing.  The typo incorrectly assigned generic contaminant values to CO2 variables.  So, for this type of mixing, both the CO2 level and the generic contaminant levels were wrong.  This has now been corrected and an existing test suite idf has been modified to use both CO2 and generic contaminant modeling.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [X] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [X] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [X] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: src/EnergyPlus/ZoneEquipmentManager.cc [hunk 1]
```diff
@@ -5950,8 +5950,8 @@ void CalcAirFlowSimple(EnergyPlusData &state,
                     state.dataContaminantBalance->MixingMassFlowCO2(ZoneB) += MassFlowToB * state.dataContaminantBalance->ZoneAirCO2(ZoneA);
                 }
                 if (state.dataContaminantBalance->Contaminant.GenericContamSimulation) {
-                    state.dataContaminantBalance->MixingMassFlowCO2(ZoneA) += MassFlowToA * state.dataContaminantBalance->ZoneAirGC(ZoneB);
-                    state.dataContaminantBalance->MixingMassFlowCO2(ZoneB) += MassFlowToB * state.dataContaminantBalance->ZoneAirGC(ZoneA);
+                    state.dataContaminantBalance->MixingMassFlowGC(ZoneA) += MassFlowToA * state.dataContaminantBalance->ZoneAirGC(ZoneB);
+                    state.dataContaminantBalance->MixingMassFlowGC(ZoneB) += MassFlowToB * state.dataContaminantBalance->ZoneAirGC(ZoneA);
```
Comment (jcyuan2020) [inline src/EnergyPlus/ZoneEquipmentManager.cc:5954]: Corrected species mass convected in to each of the zones for GC. 

Nice catch and good fixes here!

File: testfiles/RefrigeratedWarehouse.idf [hunk 2]
```diff
@@ -13319,6 +13319,28 @@
...
+    For: AllDays,            !- Field 2
+    Until: 24:00,400.;       !- Field 3
+
+  Schedule:Compact,
+    Outside GC Schedule,     !- Name
+    Any Number,              !- Schedule Type Limits Name
+    Through: 12/31,          !- Field 1
+    For: AllDays,            !- Field 2
+    Until: 24:00,23.;       !- Field 3
+
```
Comment (jcyuan2020) [inline testfiles/RefrigeratedWarehouse.idf:13343]: Constant environment background CO2 ppm and GC (ppb). OK

File: testfiles/RefrigeratedWarehouse.idf [hunk 3]
```diff
@@ -13327,6 +13349,10 @@
 
   Output:Variable,*,Zone Air Temperature,hourly;
 
+  Output:Variable,*,Zone Air CO2 Concentration,hourly;
+
+  Output:Variable,*,Zone Air Generic Air Contaminant Concentration,hourly;
```
Comment (jcyuan2020) [inline testfiles/RefrigeratedWarehouse.idf:13354]: Output variables for zone concentrations.

File: tst/EnergyPlus/unit/ZoneEquipmentManager.unit.cc [hunk 4]
```diff
@@ -62,11 +64,13 @@
 #include <EnergyPlus/DataSizing.hh>
 #include <EnergyPlus/DataZoneEnergyDemands.hh>
 #include <EnergyPlus/DataZoneEquipment.hh>
+#include <EnergyPlus/EarthTube.hh>
 #include <EnergyPlus/HeatBalanceAirManager.hh>
 #include <EnergyPlus/HeatBalanceManager.hh>
 #include <EnergyPlus/IOFiles.hh>
 #include <EnergyPlus/ScheduleManager.hh>
 #include <EnergyPlus/SimAirServingZones.hh>
+#include <EnergyPlus/ThermalChimney.hh>
```
Comment (jcyuan2020) [inline tst/EnergyPlus/unit/ZoneEquipmentManager.unit.cc:73]: I can see the need for the new .hh files inclusion here as they are need for the Manges    ManageEarthTube(), ManageCoolTower(), and ManageThermalChimney() calls.

File: tst/EnergyPlus/unit/ZoneEquipmentManager.unit.cc [hunk 5]
```diff
@@ -4406,3 +4410,78 @@ TEST_F(EnergyPlusFixture, ZoneAirMassFlowBalance_ZoneMixingInfiltrationFlowsFlag
...
+    state->dataHeatBal->RefDoorMixing(1).VolRefDoorFlowRate(1) = 1.1;
+    state->dataHeatBal->RefDoorMixing(1).VolRefDoorFlowRate(2) = 2.2;
+
+    CalcAirFlowSimple(*state);
+
+    EXPECT_NEAR(state->dataContaminantBalance->MixingMassFlowCO2(1), 395.202, 0.001);
+    EXPECT_NEAR(state->dataContaminantBalance->MixingMassFlowCO2(2), 526.884, 0.001);
+    EXPECT_NEAR(state->dataContaminantBalance->MixingMassFlowGC(1), 26.347, 0.001);
+    EXPECT_NEAR(state->dataContaminantBalance->MixingMassFlowGC(2), 13.172, 0.001);
+}
```
Comment (jcyuan2020) [inline tst/EnergyPlus/unit/ZoneEquipmentManager.unit.cc:4487]: Nice unit test!

I think it looks pretty good. I am going run some comparisons with a local pull and branch updates.

**Review by jcyuan2020 — COMMENTED** [review]
One abnormality that I noticed in the RefrigeratedWarehouse.idf is the Annual CO2 Resilience Report. The modified test files activates this Annual CO2 Resilience reporting  as Carbon Dioxide Concentration was set to `Yes' in the newly added object `ZoneAirContaminantBalance`. 

The Annual CO2 resilience report itself looks a little bit suspicious in that the second table `CO2 Level OccupantHours` shows quite a few met or unmet hours over 8760 hours (e.g the 29580.00hrs in the first cell)--and further, actually for this idf testfile the runperiod in total would be only 16 days, which is 384 hours (so should expect something even smaller than 384?): 
image

This may or may not really related to this particular fix. Possibly a new bug that may worth another investigation?

**Comment by Myoldmopar** [comment]
@jcyuan2020 just to build on your review, everything here does look good.  I built it with latest develop as well and ran the tests, everything is fine.  I am inclined to say that the hours count is out of scope for this defect and want to get this merged in.  I am going to open a new issue pointing to the comment above and merge this in.  Thanks @RKStrand and @jcyuan2020 !

**Pull request merged**